### ISSUE TITLE
Load expr import with prepared env through Dhall.TH

### ISSUE THREAD

**Issue by TristanCacqueray** [dhall-lang/dhall-haskell#2308]
In the context of a program using a dhall config, I would like to resolve `env:` code using static dhall expressions. The goal is to enable the configuration file to use `(env:APP)::{=}` where `APP` is a dhall schema shared with the program (through template haskell).

For example, this can be done using `setEnv`:

```haskell
import qualified Dhall
import qualified Dhall.TH
import System.Environment (setEnv, unsetEnv)

appPackage :: Dhall.Expr Src Void
appPackage = $(Dhall.TH.staticDhallExpression "./schemas/App.dhall")

-- | Create App :: Type
Dhall.TH.makeHaskellTypes [ Dhall.TH.SingleConstructor "App" "App" "(./schemas/App.dhall).Type" ]

loadConfig :: Text -> IO App
loadConfig = bracket_ (setEnv "APP" (Dhall.pretty appPackage)) (unsetEnv "APP") . Dhall.input Dhall.auto
```

… but that is not ideal. Is there a more efficient solution that avoids passing through the process environment?

Using `Dhall.Import.loadWith`, it seems like we should be able to cache the environment import in the initial `Status` with the expression directly (e.g. without text encoding). Unfortunately that does not seems possible. If I understand correctly, the non exported `loadImport` function would be needed, for example to implement:

```haskell
-- | Resolve all the import using an env variable map.
loadWithEnv :: [(Text, Expr Src Void)] -> Expr Src Import -> IO (Expr Src Void)
```

… In that case, would it be possible to export the missing facility? Or is there a reason to prevent such usage?

**Comment by Gabriel439** [comment]
It might be simpler to do it through a new built-in instead of the environment by following these instructions: https://hackage.haskell.org/package/dhall-1.40.1/docs/Dhall-Tutorial.html#g:23

**Comment by TristanCacqueray** [comment]
I would like to keep on using `env:APP` in the conf so that it can be evaluated outside of my program. If I understand correctly, it is not possible to write `normalizer (Embed (Import env...)) = Just appPackage` because the argument can't embed imports (e.g. ghc expects a `Expr s Void -> m (Maybe (Expr s Void))`).

**Comment by Gabriel439** [comment]
Yeah, you can't use `normalizer` for this because it only processes import-free expressions.

You can do this by hand using `traverse`, though, taking advantage of the fact that `Expr` is `Traversable` over its imports.  For example:

```haskell
traverse :: Applicative f => (a -> f b) -> Expr s a -> f (Expr s b)
```

… where `a` and `b` could be both be `Import` in this example and `f` could be `IO`:

```haskell
traverse :: (Import -> IO Import) -> Expr s Import -> IO (Expr s Import)
```

**Comment by TristanCacqueray** [comment]
I see, I didn't realized `Expr` had such `Traversable` instance. Though it seems like this would only works for the initial expression, but what if the config itself import a local file that uses the `env:APP`, is there a way to traverse further imports before they are resolved?

I think this issue can also be fixed by setting a url and its semantic hash in the environment, and then by providing a custom `Status._remote` helper to return a pretty print of the package. That way, when the cache is warm, then the package shouldn't have to be provided as text. Though I think it would be better if we could somehow by-pass the cache for such static env.

**Comment by Gabriel439** [comment]
Another approach which is more idiomatic to the language is to change the file to take the `APP` (and any other environment variables) as a function argument and then invoking the configuration as: `./config.dhall ./schemas/App.dhall`

**Comment by TristanCacqueray** [comment]
I guess another solution would be to avoid using record completions in the user config, e.g. similarly to what does `spago`, and the validation could be fully performed by the program. Though it seems a bit tricky to properly decode unions if the user config doesn't have all the variants.

**Comment by TristanCacqueray** [comment]
For what it's worth, I've implemented the env caching using a Status remote helper in [podenv/podenv@be20ef4](https://github.com/podenv/podenv/commit/be20ef44d41f00ced560e9877a7a6093587ace5c#diff-b9bc3feaca5ed3a6ac02cb7f577cb7de66aa74ab6e54faa8030f5f68ff55184dR79) and I am also adding support for weakly typed configuration using a custom Decoder in https://github.com/podenv/podenv/pull/44 .

Still I'm not sure what is the best strategy to manage such program configuration. It seems like using record completions provided by the environment would work well. Though manually decoding weakly typed config is also easier to work with.

Thank you for the suggestions @Gabriel439 , I think we can close that issue.

**Comment by TristanCacqueray** [comment]
@mmhat thank you for the suggestion, though if I understand correctly such substitutions, the configuration is no longer standalone, e.g. the language server is not usable.

**Comment by mmhat** [comment]
@TristanCacqueray Yes, indeed, that is the downside of that approach. Since I am mostly using Dhall as library for custom interpreters this is a constant source of annoyance. I did not have a look at the Dhall Language Server yet but fixing that is on my todo list.

BTW: podenv looks nice; I will look into that :+1:
